### ISSUE TITLE
Settings erased on Firefox

### ISSUE THREAD

**Issue by hackel** [uBlockOrigin/uBlock-issues#176]
### Prerequisites

- [x] I verified that this is not a filter issue
    - Filter issues MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues)
    
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uBlockOrigin](https://old.reddit.com/r/uBlockOrigin/)
    
- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
- I tried to reproduce the issue when...
    - [ ] uBlock Origin is the only extension
    - [ ] uBlock Origin with default lists/settings
    - [ ] using a new, unmodified browser profile
- [x] I am running the latest version of uBlock Origin
- [x] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### Description

Today I opened the uBlock toolbar panel to see if I needed to unblock a site, and noticed I was no longer in advanced mode.  I opened uBlock settings, and it appears that they have all been reset to the defaults.  All of my rules and filters are gone.  I've been a long-time user of uBlock for years, and never seen this happen before.

### A specific URL where the issue occurs

N/A

### Steps to Reproduce

Have not been able to reproduce it so far.

### Expected behavior:

Settings should be saved.

### Actual behavior:

Settings were erased.

### Your environment

* uBlock Origin version: 1.16.16
* Browser Name and version: 61.0.1
* Operating System and version: Ubuntu 18.04

**Comment by uBlock-user** [comment]
Is _Clear history when Firefox closes_ enabled ?

**Comment by hackel** [comment]
@uBlock-user Yes, but only for "cache" and "offline web site data."  I have not changed this setting recently, however, and it has never interfered with uBlock before.

**Comment by uBlock-user** [comment]
Duplicate of https://github.com/gorhill/uBlock/issues/3085

The bug lies in Firefox, doesn't matter which components were activated which I mentioned before, not a uBO issue.

**Comment by hackel** [comment]
@uBlock-user Maybe I'm missing something, but how is this issue related to this?  It seems to be referring to resetting only the suspendTabsUntilReady advanced preference, not erasing all settings, rules, and filters, which would be quite a critical bug in comparison.  Was that the wrong issue perhaps?  Also, that issue is quite old, yet I've been running with the current configuration for nearly a year on this profile with no problem until today.

**Comment by gorhill** [comment]
If there are no repro steps, it's unactionable, there is nothign for me to reproduce.

uBO saves its settings through `browser.storage.local`, which is a Firefox API. If for some reasons the underlying file is destroyed or corrupted, there is nothing uBO can do. It appears that for some reasons the file in which Firefox uses for `browser.storage.local` disappeared or got corrupted. You will have to investigate on your side.

**Comment by uBlock-user** [comment]
Both appear to be suffering from wiping out of data when settings in Firefox are customised and as gorhill said without STR, there's nothing to investigate in this case any further.

**Comment by uBlock-user** [comment]
> but how is this issue related to this? It seems to be referring to resetting only the suspendTabsUntilReady advanced preference, not erasing all settings, rules, and filters, which would be quite a critical bug in comparison.

https://github.com/gorhill/uBlock/issues/3085#issuecomment-333507790

Both appear to be suffering from wiping out of data and as gorhill said without STR, there's nothing to investigate in this case any further. 

Reset to default settings in Firefox and see if it still happens.

**Comment by hackel** [comment]
Yeah, fair enough, hopefully it was a one-off fluke.  Firefox should definitely have some kind of journaling in place for browser.storage.local if that's the case, but obviously not uBO's problem.  Not sure what I could do to try to trigger and reproduce this, but I'll keep an eye on it.

**Comment by hackel** [comment]
This is happening fairly consistently now, whenever my machine freezes and I have to do a hard-reset.  I even switched to a clean profile on Firefox 63.0b1 and just lost everything again.  Why is this only happening to my uBlock settings and not any other extension?  Is there something uBlock can do to mitigate it?  It is extremely frustrating and inconvenient, since there is no automatic sync/backup.

**Comment by simao** [comment]
I am having the same problem with Firefox 64.0. If my laptop crashes uBlock Origin loses all settings but other extensions are ok.

**Comment by gorhill** [comment]
Damaged files as a result of a system crash is beyond uBO's responsibility -- uBO knows nothing about files, it merely uses the browser API to save its settings. You will have to resolve you crash issues, these are the real issues.

**Comment by simao** [comment]
Indeed, I am solving those issues. I am just pointing out that other extensions can save their settings fine and ublock origin cannot, so I thought there might be something that can be done to make the storage more resilient.

**Comment by gwarser** [comment]
> I am just pointing out that other extensions can save their settings fine and ublock origin cannot

It's because uBO need to save megabytes of data. Check other extensions with huge database, or requiring lot of save/read access. Tab-Session-Manager even have wiki for this https://github.com/sienori/Tab-Session-Manager/wiki/IndexedDB-Error

**Comment by gorhill** [comment]
Well the MB of data (from cache) are saved separately from settings. Losing settings means the `browser.storage.local` got corrupted, but that is the browser dealing with file stuff, uBO just uses the API.
